dhcpv6/option_ntp_server.go

```diff
@@ -0,0 +1,163 @@
+	return fmt.Sprintf("Server Address: %s", net.IP(*n).String())
+}
+
+// NTPSuboptionMCAddr is NTP_SUBOPTION_MC_ADDR according to RFC 5908.
+type NTPSuboptionMCAddr net.IP
+
+// Code returns the suboption code.
+func (n *NTPSuboptionMCAddr) Code() OptionCode {
+	return NTPSuboptionMCAddrCode
+}
+
+// ToBytes returns the byte serialization of the suboption.
+func (n *NTPSuboptionMCAddr) ToBytes() []byte {
+	buf := uio.NewBigEndianBuffer(nil)
+	buf.Write16(uint16(NTPSuboptionMCAddrCode))
+	buf.Write16(uint16(net.IPv6len))
+	buf.WriteBytes(net.IP(*n).To16())
+	return buf.Data()
+}
+
+func (n *NTPSuboptionMCAddr) String() string {
+	return fmt.Sprintf("Multicast Address: %s", net.IP(*n).String())
+}
+
+// NTPSuboptionSrvFQDN is NTP_SUBOPTION_SRV_FQDN according to RFC 5908.
+type NTPSuboptionSrvFQDN rfc1035label.Labels
+
+// Code returns the suboption code.
+func (n *NTPSuboptionSrvFQDN) Code() OptionCode {
+	return NTPSuboptionSrvFQDNCode
+}
+
+// ToBytes returns the byte serialization of the suboption.
+func (n *NTPSuboptionSrvFQDN) ToBytes() []byte {
+	buf := uio.NewBigEndianBuffer(nil)
+	buf.Write16(uint16(NTPSuboptionSrvFQDNCode))
+	l := rfc1035label.Labels(*n)
+	buf.Write16(uint16(l.Length()))
+	buf.WriteBytes(l.ToBytes())
+	return buf.Data()
+}
+
+func (n *NTPSuboptionSrvFQDN) String() string {
+	l := rfc1035label.Labels(*n)
+	return fmt.Sprintf("Server FQDN: %s", l.String())
+}
+
+// NTPSuboptionSrvAddr is the value of NTP_SUBOPTION_SRV_ADDR according to RFC 5908.
+const (
+	NTPSuboptionSrvAddrCode = OptionCode(1)
+	NTPSuboptionMCAddrCode  = OptionCode(2)
+	NTPSuboptionSrvFQDNCode = OptionCode(3)
+)
+
+// NTPSuboptions is a collection of NTPSuboption.
+type NTPSuboptions Options
+
+// parseNTPSuboption implements the OptionParser interface.
+func parseNTPSuboption(code OptionCode, data []byte) (Option, error) {
+	//var o Options
+	buf := uio.NewBigEndianBuffer(data)
+	length := len(data)
+	data, err := buf.ReadN(length)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read %d bytes for suboption: %w", length, err)
+	}
+	switch code {
+	case NTPSuboptionSrvAddrCode, NTPSuboptionMCAddrCode:
+		if length != net.IPv6len {
+			return nil, fmt.Errorf("invalid suboption length, want %d, got %d", net.IPv6len, length)
+		}
+		var so Option
+		switch code {
+		case NTPSuboptionSrvAddrCode:
+			sos := NTPSuboptionSrvAddr(data)
+			so = &sos
+		case NTPSuboptionMCAddrCode:
+			som := NTPSuboptionMCAddr(data)
+			so = &som
+		}
+		return so, nil
+	case NTPSuboptionSrvFQDNCode:
+		l, err := rfc1035label.FromBytes(data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse rfc1035 labels: %w", err)
+		}
+		// TODO according to rfc3315, this label must not be compressed.
+		// Need to add support for compression detection to the
+		// `rfc1035label` package in order to do that.
+		so := NTPSuboptionSrvFQDN(*l)
+		return &so, nil
+	default:
+		return nil, fmt.Errorf("unknown NTP sub-option code %d", code)
```

Comment:
Maybe we should parse the rest as generic options here instead of returning an error?

Reply:
I am not sure what's right to do here. The original goal of the library was to be a scapy-like thing, where generic options and flexible parsing were first-class citizens. Then we decided that the library should do what's right for the user.

With `GenericOption` we break that a little bit, because we don't implement every possible option, and there are custom vendor options too, so we don't have a choice.
With the NTP Server option we know that only three sub-options are supported by the standard, and I think we should error out. If at any point a new option is added, I think we can reactively add support to it.

Thoughts?

Reply:
I'd rather not be that strict and use the same approach we use for the [main options parsing](https://github.com/insomniacslk/dhcp/blob/master/dhcpv6/options.go#L107).

In the same way we parse "main" options we know nothing about, I think that approach should be followed here.

If a new sub option is added in the future, this will break the parsing of the whole packet which is not nice (vs reactively adding support for it).